rust-lib/src/lib.rs

```diff
@@ -120,36 +125,64 @@ where
     }
 }
 
-#[derive(Debug, Deserialize)]
-struct StartRgbArgs {
-    #[serde(with = "serde_with::rust::display_fromstr")]
-    network: bp::Chain,
-    #[serde(with = "serde_with::rust::display_fromstr")]
-    stash_endpoint: SocketLocator,
-    contract_endpoints: HashMap<ContractName, String>,
-    threaded: bool,
-    datadir: String,
+#[derive(Debug, Display, From, Error)]
+#[display(doc_comments)]
+#[non_exhaustive]
+enum RequestError {
+    /// String error
+    #[from]
+    StringErr(String),
+
+    /// Input value is not a JSON object
+    #[from]
+    SerdeErr(serde_json::Error),
+
+    /// Invoice error
+    #[from]
+    InvoiceErr(rgb::fungible::InvoiceError),
+
+    /// Input value is not a string
+    #[from]
+    Utf8Err(std::str::Utf8Error),
+
+    /// Invalid network
+    #[from]
+    ChainPaseErr(rgb::lnpbp::bp::chain::ParseError),
 }
 
-fn _start_rgb(json: *mut c_char) -> Result<Runtime, String> {
-    let config: StartRgbArgs =
-        serde_json::from_str(ptr_to_string(json)?.as_str()).map_err(|e| format!("{:?}", e))?;
-    info!("Config: {:?}", config);
+fn _start_rgb(network: *mut c_char, stash_endpoint: *mut c_char, contract_endpoints: *mut c_char,
+              threaded: bool, datadir: *mut c_char) -> Result<Runtime, RequestError> {
+
+    let c_network = unsafe { CStr::from_ptr(network) };
+    let network = bp::Chain::from_str(c_network.to_str()?)?;
+
+    let c_stash_endpoint = unsafe { CStr::from_ptr(stash_endpoint) };
+    let stash_endpoint =
+        SocketLocator::Posix(path::PathBuf::from(c_stash_endpoint.to_str()?.to_string()));
+
+    let contract_endpoints: HashMap<ContractName, String> =
+        serde_json::from_str(ptr_to_string(contract_endpoints)?.as_str()).map_err(|e| format!("{:?}", e))?;
```

Comment:
We do not need to map already supported JSON error into string representation - plus we can use simple reference operator `&` instead of `.as_str()`, which will increase code readability
```suggestion
        serde_json::from_str(&ptr_to_string(contract_endpoints)?)?;
```

Reply:
good to know, thanks! done